File: lib/onr_bdlight_xml.py
"""Validação do XML BANCOLIGHT (BD Light) — modelo spec § 4.1 / bdlight-exemplo-1.xml."""

from __future__ import annotations

import re
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.request import urlopen
from xml.etree import ElementTree as ET

ROOT_TAG = "BANCOLIGHT"
INDIVIDUO_TAG = "INDIVIDUO"
REQUIRED_FIELDS = (
    "NOME",
    "CNPJCPF",
    "NMATRICULA",
    "TIPODEATO",
    "DTREGAVERB",
    "DTVENDA",
)
MAX_BYTES = 5 * 1024 * 1024
DATE_RE = re.compile(r"^\d{8}$")
CPF_CNPJ_RE = re.compile(r"^\d{11}$|^\d{14}$")
FETCH_TIMEOUT_SEC = 60


class BdlightXmlValidationError(Exception):
    """XML fora do padrão BANCOLIGHT."""

    def __init__(self, source: str, errors: list[str]):
        self.source = source
        self.errors = errors
        detail = "\n".join(f"  - {e}" for e in errors)
        super().__init__(f"XML BD Light inválido ({source}):\n{detail}")


def _validation_enabled() -> bool:
    import os

    raw = (os.getenv("BDLIGHT_IMPORTAR_VALIDAR_XML") or "true").strip().lower()
    return raw in ("1", "true", "yes")


def _decode_xml(content: bytes) -> str:
    if len(content) > MAX_BYTES:
        raise BdlightXmlValidationError(
            "conteúdo",
            [f"Tamanho {len(content)} bytes excede o máximo de {MAX_BYTES} (5MB)."],
        )
    for encoding in ("utf-8", "iso-8859-1", "latin-1"):
        try:
            return content.decode(encoding)
        except UnicodeDecodeError:
            continue
    raise BdlightXmlValidationError(
        "conteúdo", ["Não foi possível decodificar o XML (utf-8 / iso-8859-1)."]
    )


def _element_text(parent: ET.Element, tag: str) -> str:
    child = parent.find(tag)
    if child is None:
        return ""
    return "".join(child.itertext()).strip()


def _validate_date(value: str, field: str, *, required: bool) -> list[str]:
    if not value:
        return [] if not required else [f"{field} é obrigatório."]
    if not DATE_RE.match(value):
        return [f"{field} deve ter 8 dígitos (DDMMAAAA), recebido: {value!r}."]
    return []


def _validate_individuo(node: ET.Element, index: int) -> list[str]:
    errors: list[str] = []
    label = f"INDIVIDUO[{index}]"

    for field in REQUIRED_FIELDS:
        if node.find(field) is None:
            errors.append(f"{label}: campo obrigatório ausente: {field}.")

    nome = _element_text(node, "NOME")
    if not nome:
        errors.append(f"{label}: NOME não pode ser vazio.")

    cpf_cnpj = re.sub(r"\D", "", _element_text(node, "CNPJCPF"))
    if not CPF_CNPJ_RE.match(cpf_cnpj):
        errors.append(
            f"{label}: CNPJCPF deve ter 11 (CPF) ou 14 (CNPJ) dígitos, "
            f"recebido: {_element_text(node, 'CNPJCPF')!r}."
        )

    matricula = _element_text(node, "NMATRICULA")
    if not matricula:
        errors.append(f"{label}: NMATRICULA não pode ser vazio.")

    tipo = _element_text(node, "TIPODEATO")
    if not tipo:
        errors.append(f"{label}: TIPODEATO não pode ser vazio.")

    errors.extend(
        _validate_date(_element_text(node, "DTREGAVERB"), f"{label}.DTREGAVERB", required=True)
    )
    dtvenda = _element_text(node, "DTVENDA")
    if dtvenda:
        errors.extend(
            _validate_date(dtvenda, f"{label}.DTVENDA", required=False)
        )

    allowed = set(REQUIRED_FIELDS)
    for child in node:
        tag = child.tag.split("}")[-1] if "}" in child.tag else child.tag
        if tag not in allowed:
            errors.append(f"{label}: elemento não esperado: {tag}.")

    return errors


def validate_bdlight_xml(content: bytes | str, *, source: str = "XML") -> dict:
    """
    Valida estrutura BANCOLIGHT/INDIVIDUO conforme bdlight-exemplo-1.xml.

    Retorna dict com valid=True e quantidade de INDIVIDUO.
    Levanta BdlightXmlValidationError se inválido.
    """
    raw = content.encode("utf-8") if isinstance(content, str) else content
    text = _decode_xml(raw).strip()
    if not text:
        raise BdlightXmlValidationError(source, ["Arquivo XML vazio."])

    errors: list[str] = []
    try:
        root = ET.fromstring(text)
    except ET.ParseError as exc:
        raise BdlightXmlValidationError(source, [f"XML malformado: {exc}"]) from exc

    tag = root.tag.split("}")[-1] if "}" in root.tag else root.tag
    if tag != ROOT_TAG:
        errors.append(
            f"Elemento raiz deve ser {ROOT_TAG}, encontrado: {tag!r}."
        )

    individuos = [
        el
        for el in root
        if (el.tag.split("}")[-1] if "}" in el.tag else el.tag) == INDIVIDUO_TAG
    ]
    if not individuos:
        errors.append(f"É necessário ao menos um {INDIVIDUO_TAG}.")

    for i, node in enumerate(individuos, start=1):
        errors.extend(_validate_individuo(node, i))

    if errors:
        raise BdlightXmlValidationError(source, errors)

    return {"valid": True, "individuos": len(individuos), "source": source}


def validate_bdlight_xml_file(path: str | Path) -> dict:
    p = Path(path)
    if not p.is_file():
        raise BdlightXmlValidationError(str(p), ["Arquivo não encontrado."])
    return validate_bdlight_xml(p.read_bytes(), source=str(p))


def validate_bdlight_xml_url(url: str) -> dict:
    try:
        with urlopen(url, timeout=FETCH_TIMEOUT_SEC) as resp:
            content = resp.read()
    except HTTPError as exc:
        raise BdlightXmlValidationError(
            url, [f"HTTP {exc.code} ao baixar URL."]
        ) from exc
    except URLError as exc:
        raise BdlightXmlValidationError(
            url, [f"Erro ao baixar URL: {exc.reason}"]
        ) from exc
    return validate_bdlight_xml(content, source=url)


def assert_bdlight_xml_valid(content: bytes | str, *, source: str = "XML") -> dict:
    return validate_bdlight_xml(content, source=source)


def load_xml_paths_from_env() -> list[str]:
    import os

    raw = (os.getenv("BDLIGHT_IMPORTAR_XML_PATH") or "").strip()
    if not raw:
        return []
    return [p for p in re.split(r"[,;]+", raw) if p.strip()]


def validate_import_inputs(
    arquivos: list[dict[str, str]],
    *,
    xml_paths: list[str] | None = None,
) -> list[dict]:
    """Valida paths locais e URLs antes do ImportarArquivoBDL."""
    if not _validation_enabled():
        return []

    results: list[dict] = []
    paths = xml_paths if xml_paths is not None else load_xml_paths_from_env()

    for path in paths:
        results.append(validate_bdlight_xml_file(path.strip()))

    for row in arquivos:
        url = row.get("URLArquivo", "")
        if url:
            results.append(validate_bdlight_xml_url(url))

    return results

File: lib/onr_bdlight_xml.js
/**
 * Validação do XML BANCOLIGHT (BD Light) — modelo spec § 4.1 / bdlight-exemplo-1.xml.
 */
import fs from "node:fs";

export const ROOT_TAG = "BANCOLIGHT";
export const INDIVIDUO_TAG = "INDIVIDUO";
export const REQUIRED_FIELDS = [
  "NOME",
  "CNPJCPF",
  "NMATRICULA",
  "TIPODEATO",
  "DTREGAVERB",
  "DTVENDA",
];
export const MAX_BYTES = 5 * 1024 * 1024;
const DATE_RE = /^\d{8}$/;
const CPF_CNPJ_RE = /^\d{11}$|^\d{14}$/;
const FETCH_TIMEOUT_MS = 60_000;

export class BdlightXmlValidationError extends Error {
  constructor(source, errors) {
    const detail = errors.map((e) => `  - ${e}`).join("\n");
    super(`XML BD Light inválido (${source}):\n${detail}`);
    this.name = "BdlightXmlValidationError";
    this.source = source;
    this.errors = errors;
  }
}

function validationEnabled() {
  const raw = (process.env.BDLIGHT_IMPORTAR_VALIDAR_XML || "true").trim().toLowerCase();
  return ["1", "true", "yes"].includes(raw);
}

function decodeXml(buffer) {
  if (buffer.length > MAX_BYTES) {
    throw new BdlightXmlValidationError("conteúdo", [
      `Tamanho ${buffer.length} bytes excede o máximo de ${MAX_BYTES} (5MB).`,
    ]);
  }
  for (const encoding of ["utf8", "latin1"]) {
    try {
      return buffer.toString(encoding);
    } catch {
      /* try next */
    }
  }
  throw new BdlightXmlValidationError("conteúdo", [
    "Não foi possível decodificar o XML (utf-8 / iso-8859-1).",
  ]);
}

function stripCdata(text) {
  return text.replace(/<!\[CDATA\[([\s\S]*?)]]>/g, "$1").trim();
}

function tagText(block, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)</${tag}>`, "i");
  const match = block.match(re);
  if (!match) return "";
  return stripCdata(match[1]).trim();
}

function validateDate(value, field, { required }) {
  if (!value) {
    return required ? [`${field} é obrigatório.`] : [];
  }
  if (!DATE_RE.test(value)) {
    return [`${field} deve ter 8 dígitos (DDMMAAAA), recebido: ${JSON.stringify(value)}.`];
  }
  return [];
}

function extractIndividuoBlocks(xml) {
  const re = /<INDIVIDUO\b[^>]*>([\s\S]*?)<\/INDIVIDUO>/gi;
  const blocks = [];
  let match;
  while ((match = re.exec(xml)) !== null) {
    blocks.push(match[0]);
  }
  return blocks;
}

function validateIndividuo(block, index) {
  const errors = [];
  const label = `INDIVIDUO[${index}]`;

  for (const field of REQUIRED_FIELDS) {
    if (!new RegExp(`<${field}\\b`, "i").test(block)) {
      errors.push(`${label}: campo obrigatório ausente: ${field}.`);
    }
  }

  const nome = tagText(block, "NOME");
  if (!nome) errors.push(`${label}: NOME não pode ser vazio.`);

  const cpfRaw = tagText(block, "CNPJCPF");
  const cpfDigits = cpfRaw.replace(/\D/g, "");
  if (!CPF_CNPJ_RE.test(cpfDigits)) {
    errors.push(
      `${label}: CNPJCPF deve ter 11 (CPF) ou 14 (CNPJ) dígitos, recebido: ${JSON.stringify(cpfRaw)}.`
    );
  }

  if (!tagText(block, "NMATRICULA")) {
    errors.push(`${label}: NMATRICULA não pode ser vazio.`);
  }
  if (!tagText(block, "TIPODEATO")) {
    errors.push(`${label}: TIPODEATO não pode ser vazio.`);
  }

  errors.push(
    ...validateDate(tagText(block, "DTREGAVERB"), `${label}.DTREGAVERB`, { required: true })
  );
  const dtvenda = tagText(block, "DTVENDA");
  if (dtvenda) {
    errors.push(...validateDate(dtvenda, `${label}.DTVENDA`, { required: false }));
  }

  return errors;
}

/**
 * Valida estrutura BANCOLIGHT/INDIVIDUO conforme bdlight-exemplo-1.xml.
 * @param {Buffer|string} content
 * @param {{ source?: string }} [opts]
 */
export function validateBdlightXml(content, { source = "XML" } = {}) {
  const buffer = Buffer.isBuffer(content)
    ? content
    : Buffer.from(String(content), "utf8");
  const text = decodeXml(buffer).trim();
  if (!text) {
    throw new BdlightXmlValidationError(source, ["Arquivo XML vazio."]);
  }

  const errors = [];
  if (!/<BANCOLIGHT\b/i.test(text)) {
    errors.push(`Elemento raiz deve ser ${ROOT_TAG}.`);
  }

  const blocks = extractIndividuoBlocks(text);
  if (!blocks.length) {
    errors.push(`É necessário ao menos um ${INDIVIDUO_TAG}.`);
  }

  blocks.forEach((block, i) => {
    errors.push(...validateIndividuo(block, i + 1));
  });

  if (errors.length) {
    throw new BdlightXmlValidationError(source, errors);
  }

  return { valid: true, individuos: blocks.length, source };
}

export function validateBdlightXmlFile(filePath) {
  if (!fs.existsSync(filePath)) {
    throw new BdlightXmlValidationError(filePath, ["Arquivo não encontrado."]);
  }
  return validateBdlightXml(fs.readFileSync(filePath), { source: filePath });
}

export async function validateBdlightXmlUrl(url) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), FETCH_TIMEOUT_MS);
  let response;
  try {
    response = await fetch(url, { signal: controller.signal });
  } catch (err) {
    const msg =
      err.name === "AbortError"
        ? "Timeout ao baixar URL."
        : `Erro ao baixar URL: ${err.message}`;
    throw new BdlightXmlValidationError(url, [msg]);
  } finally {
    clearTimeout(timer);
  }

  if (!response.ok) {
    throw new BdlightXmlValidationError(url, [
      `HTTP ${response.status} ao baixar URL.`,
    ]);
  }

  const buffer = Buffer.from(await response.arrayBuffer());
  return validateBdlightXml(buffer, { source: url });
}

export function loadXmlPathsFromEnv() {
  const raw = (process.env.BDLIGHT_IMPORTAR_XML_PATH || "").trim();
  if (!raw) return [];
  return raw
    .split(/[,;]+/)
    .map((p) => p.trim())
    .filter(Boolean);
}

/**
 * Valida paths locais (BDLIGHT_IMPORTAR_XML_PATH) e URLs do envelope.
 */
export async function validateImportInputs(arquivos, { xmlPaths } = {}) {
  if (!validationEnabled()) return [];

  const results = [];
  const paths = xmlPaths ?? loadXmlPathsFromEnv();

  for (const path of paths) {
    results.push(validateBdlightXmlFile(path));
  }

  for (const row of arquivos) {
    const url = row.URLArquivo;
    if (url) {
      results.push(await validateBdlightXmlUrl(url));
    }
  }

  return results;
}
